src/Compilers/CSharp/Portable/Lowering/RuntimeAsyncIteratorRewriter/RuntimeAsyncIteratorRewriter.MoveNextAsyncRewriter.cs

```diff
@@ -0,0 +1,462 @@
+
+                blockBuilder.Add(GenerateAllHoistedLocalsCleanup());
+
+                // if (this.combinedTokens != null) { this.combinedTokens.Dispose(); this.combinedTokens = null; }
+                addDisposeCombinedTokensIfNeeded(blockBuilder);
+
+                // _current = default;
+                blockBuilder.Add(GenerateClearCurrent());
+
+                // throw null;
+                blockBuilder.Add(F.Throw(null));
+
+                return new BoundCatchBlock(
+                    F.Syntax,
+                    [exceptionLocal],
+                    F.Local(exceptionLocal),
+                    exceptionLocal.Type,
+                    exceptionFilterPrologueOpt: null,
+                    exceptionFilterOpt: null,
+                    body: F.Block(blockBuilder.ToImmutableAndFree()),
+                    isSynthesizedAsyncCatchAll: true);
+            }
+
+            void addDisposeCombinedTokensIfNeeded(ArrayBuilder<BoundStatement> builder)
+            {
+                AsyncIteratorMethodToStateMachineRewriter.AddDisposeCombinedTokensIfNeeded(builder, _combinedTokensField, F);
+            }
+
+            // Lower the body, adding an entry state (-3) at the start,
+            // so that we can differentiate an async-iterator that was never moved forward with MoveNextAsync()
+            // from one that is running (-1).
+            // Then we can guard against some bad usages of DisposeAsync.
+            BoundStatement visitBody(BoundStatement body)
+            {
+                // Produce:
+                //  initialStateResumeLabel:
+                //  if (disposeMode) goto _exprReturnLabel;
+                //  this.state = cachedState = -1;
+                //  ... rewritten body, which ends with `disposeMode = true; goto topLevelDisposeLabel;` ...
+
+                AddState(StateMachineState.InitialAsyncIteratorState, out GeneratedLabelSymbol resumeLabel);
+
+                var rewrittenBody = (BoundStatement)Visit(body);
+
+                return F.Block(
+                    F.Label(resumeLabel), // initialStateResumeLabel:
+                    GenerateConditionalJumpToCurrentDisposalLabel(), // if (disposeMode) goto _exprReturnLabel;
+                    GenerateSetBothStates(StateMachineState.NotStartedOrRunningState), // this.state = cachedState = -1;
+                    rewrittenBody);
+            }
+        }
+
+        internal BoundStatement GenerateTopLevelTry(BoundBlock tryBlock, ImmutableArray<BoundCatchBlock> catchBlocks)
+            => F.Try(tryBlock, catchBlocks);
+
+        private BoundExpressionStatement GenerateClearCurrent()
+        {
+            Debug.Assert(_currentField is not null);
+
+            // _current = default;
+            return F.Assignment(F.InstanceField(_currentField), F.Default(_currentField.Type));
+        }
+
+        private BoundStatement GenerateConditionalJumpToCurrentDisposalLabel()
+        {
+            Debug.Assert(_currentDisposalLabel is object);
```

Comment:
Nit: the file mixes `is object` and `is not null` checks. Consider deciding on just one style, as this is new code.